### DISCUSSION TITLE
关于在世纪互联下，图片文件预览图的想法。

### DISCUSSION THREAD

**Discussion by Android-KitKat** [spencerwooo/onedrive-vercel-index#378] [Ideas]
## 情况
我用的世纪互联的 OneDrive。
在更新到最新提交 920e83e 后，网格视图下图片文件的显示好像有些问题。
使用自动命名的话，完全分不清是哪些图片。
![相册](https://user-images.githubusercontent.com/6592315/152987846-028da643-b834-43fe-9e3a-c17bfe70aa86.png)

## 原因
[检索文件或文件夹的缩略图](https://docs.microsoft.com/zh-cn/graph/api/driveitem-list-thumbnails?view=graph-rest-1.0&tabs=http#remarks)
我翻了翻文档，发现好像 OneDrive for Business 和 SharePoint 不支持批量获取缩略图。
(但现在好像只有世纪互联的不行？)
我试过单独获取缩略图，确实是可以的。

## 想法
如果一个一个单独获取的话……
感觉好像很容易执行超时。

或许可以在配置中提供一个替代选项？
启用后，对于图片文件，直接将 `downloadUrl` 也就是 `raw=true` 返回的地址作为预览图显示。
而文本、演示文稿、PDF、视频等其他文件的预览图好像就没办法了。

**Comment by myl7** [comment]
原图（也就是用 `raw=true`）应该是没法做预览图的，一方面占用大量带宽，另一方面 OneDrive 也有 ratelimit

一个个单独获取倒是有可能，例如搜索的结果就是一个个获取的（如果是非个人版 OneDrive 的话）

**Reply by Android-KitKat** [reply]
有道理。
其实我这样想，是因为之前项目还用 `react-viewer` 的时候，就是用的原图作为下方的预览列表。

我明白大佬的意思了，确实可以一个个获取。
你是指，并不是在列出文件列表时，单次访问中去一个个获取(那样肯定会超时)。
而是类似于 `raw=true` 这样，新增一个 API 专门用来获取文件的缩略图。
让浏览器自己一个个分别请求。

而且这样的话，可能其他非图片文件的缩略图也没问题了？
(我试了下，docx、xlsx、pptx、pdf 也有缩略图。)

**Comment by spencerwooo** [comment]
我问一下，E5 Developer Subscription 版本不是 OneDrive for Business 吗？Demo 目前是 E5，我看缩略图是正常请求的啊？

**Reply by Android-KitKat** [reply]
E5 应该算 OneDrive for Business 吧……
难道现在只有世纪互联的不能用 `expand=thumbnails`。
我的部署地址：[Android 的公开网盘](https://drive.android99.me)

实际上我在官方的 OneDrive 网页版上也看不了缩略图。
(但用 `/me/drive/root:/path/to/file:/thumbnails` 这个 API 可以。)
怀疑是不是管理员设置有什么问题，还去查了下 [Set-SPOTenant (SharePointOnlinePowerShell)](https://docs.microsoft.com/zh-cn/powershell/module/sharepoint-online/set-spotenant?view=sharepoint-ps)，不过没看到有关系的设置。

**Comment by myl7** [comment]
Ref #387
我也没有世纪互联可供测试，可能需要你自己 merge 了测试一下

**Reply by Android-KitKat** [reply]
部署上去了，用的 myl7/onedrive-vercel-index@197336d 提交。
[Android 的公开网盘](https://drive.android99.me)
不过好像没什么效果。
`/api/thumbnail/?path=string` 是可以用的。

**Reply by myl7** [reply]
我又 push 了一下，你再试试？

**Reply by Android-KitKat** [reply]
好耶，有缩略图了。
大佬辛苦了！
应该没问题了，可以去看看我的部署。
![相册](https://user-images.githubusercontent.com/6592315/153155209-9cda8ddf-a04c-41a2-83a4-2dc1a02e029e.png)

**Comment by Android-KitKat** [comment]
引用 #387
> For authorization, since thumbnails are related to file content, the new API takes path as arg, and check if it is under any protected route. Currently protected pathes are simply rejected (because the url is passed directly to <img>, which can not customize headers). https://stackoverflow.com/questions/34096744/how-should-i-load-images-if-i-use-token-based-authentication may be a solution though I am not sure whether it is worthy.

|柱|ω・`) 嗯……受保护的路径，我也想不到怎么解决比较好。
~~虽然我目前没有设置受保护的路径。~~
